File: algorithms/python/guesstheWord/guesstheWord.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/guess-the-word/
# Author: Miao Zhang
# Date:   2021-03-17

# """
# This is Master's API interface.
# You should not implement it, or speculate about its implementation
# """
# class Master:
#     def guess(self, word: str) -> int:

class Solution:
    def findSecretWord(self, wordlist: List[str], master: 'Master') -> None:
        
        def match(a, b):
            res = 0
            for i in range(len(a)):
                if a[i] == b[i]: res += 1
            return res
        
        cnt = 0
        i = 0
        while i < 10 and cnt < 6:
            guess = wordlist[random.randint(0, len(wordlist) - 1)]
            cnt = master.guess(guess)
            words = []
            for w in wordlist:
                if match(guess, w) == cnt:
                    words.append(w)
            wordlist = words
            i += 1
        

File: algorithms/cpp/guesstheWord/guesstheWord.cpp
// Source: https://leetcode.com/problems/guess-the-word/
// Author: Miao Zhang
// Date:   2021-03-17

/**
 * // This is the Master's API interface.
 * // You should not implement it, or speculate about its implementation
 * class Master {
 *   public:
 *     int guess(string word);
 * };
 */
class Solution {
public:
    void findSecretWord(vector<string>& wordlist, Master& master) {
        for (int i = 0, cnt = 0; i < 10 && cnt < 6; i++) {
            string guess = wordlist[rand() % wordlist.size()];
            cnt = master.guess(guess);
            vector<string> words;
            for (string& w: wordlist) {
                if (match(guess, w) == cnt) {
                    words.push_back(w);
                }
            }
            wordlist = words;
        }
    }

    int match(string& a, string& b) {
        int res = 0;
        for (int i = 0; i < a.size(); i++) {
            if (a[i] == b[i]) res++;
        }
        return res;
    }
};
